backend/api/src/api/v1/gql/mod.rs

```diff
@@ -0,0 +1,142 @@
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use async_graphql::{
+    extensions, futures_util::Stream, ComplexObject, Context, Schema, SimpleObject, Subscription,
+};
+use common::types::{session, user};
+use hyper::{Body, Response};
+use routerify::Router;
+
+use crate::{api::error::RouteError, global::GlobalState};
+
+use self::error::{GqlError, Result, ResultExt};
+
+pub mod auth;
+pub mod error;
+pub mod handlers;
+pub mod models;
+
+#[derive(Default)]
+pub struct GqlContext {
+    pub is_websocket: bool,
+    pub session: ArcSwap<Option<session::Model>>,
+}
+
+impl GqlContext {
+    pub async fn get_session(&self, global: &Arc<GlobalState>) -> Result<Option<session::Model>> {
+        let guard = self.session.load();
+        let Some(session) = guard.as_ref() else {
+            return Ok(None)
+        };
+
+        if !self.is_websocket {
+            if !session.validate() {
+                return Err(GqlError::InvalidSession.with_message("Session is no longer valid"));
+            }
+
+            return Ok(Some(session.clone()));
+        }
+
+        let session = sqlx::query_as!(
+            session::Model,
+            "SELECT * FROM sessions WHERE id = $1",
+            session.id
+        )
+        .fetch_optional(&global.db)
+        .await
+        .extend_gql("failed to fetch session")?
```

Comment:
Same concept as the above route error. Perhaps a better function name. I took the idea of `extend_` from the GQL library, since they use that prefix to add their custom error.